In [55]:
import pandas as pd
import numpy as np
from sklearn import preprocessing
import matplotlib.pyplot as plt 
plt.rc("font", size=14)
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import seaborn as sns
sns.set(style="white")
sns.set(style="whitegrid", color_codes=True)

In [57]:
df = pd.read_csv(r'C:\Users\Dell\Downloads\avocado.csv')
df

Unnamed: 0        Date  AveragePrice  Total Volume     4046       4225  \
0             0.0  27-12-2015          1.33      64236.62  1036.74   54454.85   
1             1.0  20-12-2015          1.35      54876.98   674.28   44638.81   
2             2.0  13-12-2015          0.93     118220.22   794.70  109149.67   
3             3.0  06-12-2015          1.08      78992.15  1132.00   71976.41   
4             4.0  29-11-2015          1.28      51039.60   941.48   43838.39   
...           ...         ...           ...           ...      ...        ...   
16463         NaN         NaN           NaN           NaN      NaN        NaN   
16464         NaN         NaN           NaN           NaN      NaN        NaN   
16465         NaN         NaN           NaN           NaN      NaN        NaN   
16466         NaN         NaN           NaN           NaN      NaN        NaN   
16467         NaN         NaN           NaN           NaN      NaN        NaN   

         4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0       48.16     8696.87     8603.62       93.25          0.0  conventional   
1       58.33     9505.56     9408.07       97.49          0.0  conventional   
2      130.50     8145.35     8042.21      103.14          0.0  conventional   
3       72.58     5811.16     5677.40      133.76          0.0  conventional   
4       75.78     6183.95     5986.26      197.69          0.0  conventional   
...       ...         ...         ...         ...          ...           ...   
16463     NaN         NaN         NaN         NaN          NaN           NaN   
16464     NaN         NaN         NaN         NaN          NaN           NaN   
16465     NaN         NaN         NaN         NaN          NaN           NaN   
16466     NaN         NaN         NaN         NaN          NaN           NaN   
16467     NaN         NaN         NaN         NaN          NaN           NaN   

         year  region  
0      2015.0  Albany  
1      2015.0  Albany  
2      2015.0  Albany  
3      2015.0  Albany  
4      2015.0  Albany  
...       ...     ...  
16463     NaN     NaN  
16464     NaN     NaN  
16465     NaN     NaN  
16466     NaN     NaN  
16467     NaN     NaN  

[16468 rows x 14 columns]

In [58]:
df['Date'] = pd.to_datetime(df['Date'])
df

C:\Users\Dell\AppData\Local\Temp\ipykernel_10960\66253117.py:1: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  df['Date'] = pd.to_datetime(df['Date'])


Unnamed: 0       Date  AveragePrice  Total Volume     4046       4225  \
0             0.0 2015-12-27          1.33      64236.62  1036.74   54454.85   
1             1.0 2015-12-20          1.35      54876.98   674.28   44638.81   
2             2.0 2015-12-13          0.93     118220.22   794.70  109149.67   
3             3.0 2015-06-12          1.08      78992.15  1132.00   71976.41   
4             4.0 2015-11-29          1.28      51039.60   941.48   43838.39   
...           ...        ...           ...           ...      ...        ...   
16463         NaN        NaT           NaN           NaN      NaN        NaN   
16464         NaN        NaT           NaN           NaN      NaN        NaN   
16465         NaN        NaT           NaN           NaN      NaN        NaN   
16466         NaN        NaT           NaN           NaN      NaN        NaN   
16467         NaN        NaT           NaN           NaN      NaN        NaN   

         4770  Total Bags  Small Bags  Large Bags  XLarge Bags          type  \
0       48.16     8696.87     8603.62       93.25          0.0  conventional   
1       58.33     9505.56     9408.07       97.49          0.0  conventional   
2      130.50     8145.35     8042.21      103.14          0.0  conventional   
3       72.58     5811.16     5677.40      133.76          0.0  conventional   
4       75.78     6183.95     5986.26      197.69          0.0  conventional   
...       ...         ...         ...         ...          ...           ...   
16463     NaN         NaN         NaN         NaN          NaN           NaN   
16464     NaN         NaN         NaN         NaN          NaN           NaN   
16465     NaN         NaN         NaN         NaN          NaN           NaN   
16466     NaN         NaN         NaN         NaN          NaN           NaN   
16467     NaN         NaN         NaN         NaN          NaN           NaN   

         year  region  
0      2015.0  Albany  
1      2015.0  Albany  
2      2015.0  Albany  
3      2015.0  Albany  
4      2015.0  Albany  
...       ...     ...  
16463     NaN     NaN  
16464     NaN     NaN  
16465     NaN     NaN  
16466     NaN     NaN  
16467     NaN     NaN  

[16468 rows x 14 columns]

In [59]:
#df.drop(['4046', '4225', '4770'], axis=1, inplace=True)
df.drop(['Date',], axis=1, inplace=True)

In [60]:
df

Unnamed: 0  AveragePrice  Total Volume     4046       4225    4770  \
0             0.0          1.33      64236.62  1036.74   54454.85   48.16   
1             1.0          1.35      54876.98   674.28   44638.81   58.33   
2             2.0          0.93     118220.22   794.70  109149.67  130.50   
3             3.0          1.08      78992.15  1132.00   71976.41   72.58   
4             4.0          1.28      51039.60   941.48   43838.39   75.78   
...           ...           ...           ...      ...        ...     ...   
16463         NaN           NaN           NaN      NaN        NaN     NaN   
16464         NaN           NaN           NaN      NaN        NaN     NaN   
16465         NaN           NaN           NaN      NaN        NaN     NaN   
16466         NaN           NaN           NaN      NaN        NaN     NaN   
16467         NaN           NaN           NaN      NaN        NaN     NaN   

       Total Bags  Small Bags  Large Bags  XLarge Bags          type    year  \
0         8696.87     8603.62       93.25          0.0  conventional  2015.0   
1         9505.56     9408.07       97.49          0.0  conventional  2015.0   
2         8145.35     8042.21      103.14          0.0  conventional  2015.0   
3         5811.16     5677.40      133.76          0.0  conventional  2015.0   
4         6183.95     5986.26      197.69          0.0  conventional  2015.0   
...           ...         ...         ...          ...           ...     ...   
16463         NaN         NaN         NaN          NaN           NaN     NaN   
16464         NaN         NaN         NaN          NaN           NaN     NaN   
16465         NaN         NaN         NaN          NaN           NaN     NaN   
16466         NaN         NaN         NaN          NaN           NaN     NaN   
16467         NaN         NaN         NaN          NaN           NaN     NaN   

       region  
0      Albany  
1      Albany  
2      Albany  
3      Albany  
4      Albany  
...       ...  
16463     NaN  
16464     NaN  
16465     NaN  
16466     NaN  
16467     NaN  

[16468 rows x 13 columns]

In [61]:
df.isna().sum()

Unnamed: 0      14951
AveragePrice    14951
Total Volume    14951
4046            14951
4225            14951
4770            14951
Total Bags      14951
Small Bags      14951
Large Bags      14951
XLarge Bags     14951
type            14951
year            14951
region          14951
dtype: int64

In [62]:
df.fillna(df.mean(), inplace=True)

C:\Users\Dell\AppData\Local\Temp\ipykernel_10960\820435583.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.fillna(df.mean(), inplace=True)


In [63]:
df.isna().sum()

Unnamed: 0          0
AveragePrice        0
Total Volume        0
4046                0
4225                0
4770                0
Total Bags          0
Small Bags          0
Large Bags          0
XLarge Bags         0
type            14951
year                0
region          14951
dtype: int64

In [64]:
from sklearn.preprocessing import LabelEncoder

columns_to_encode = ['type', 'region']  
le = LabelEncoder()

for col in columns_to_encode:
    df[col] = le.fit_transform(df[col])

In [65]:
df

Unnamed: 0  AveragePrice  Total Volume           4046           4225  \
0        0.000000       1.33000  6.423662e+04    1036.740000   54454.850000   
1        1.000000       1.35000  5.487698e+04     674.280000   44638.810000   
2        2.000000       0.93000  1.182202e+05     794.700000  109149.670000   
3        3.000000       1.08000  7.899215e+04    1132.000000   71976.410000   
4        4.000000       1.28000  5.103960e+04     941.480000   43838.390000   
...           ...           ...           ...            ...            ...   
16463   26.995386       1.07499  1.601879e+06  646438.654113  611437.502591   
16464   26.995386       1.07499  1.601879e+06  646438.654113  611437.502591   
16465   26.995386       1.07499  1.601879e+06  646438.654113  611437.502591   
16466   26.995386       1.07499  1.601879e+06  646438.654113  611437.502591   
16467   26.995386       1.07499  1.601879e+06  646438.654113  611437.502591   

               4770     Total Bags     Small Bags    Large Bags  XLarge Bags  \
0         48.160000    8696.870000    8603.620000     93.250000     0.000000   
1         58.330000    9505.560000    9408.070000     97.490000     0.000000   
2        130.500000    8145.350000    8042.210000    103.140000     0.000000   
3         72.580000    5811.160000    5677.400000    133.760000     0.000000   
4         75.780000    6183.950000    5986.260000    197.690000     0.000000   
...             ...            ...            ...           ...          ...   
16463  50405.495399  293597.415735  248773.595287  42642.049374  2181.771074   
16464  50405.495399  293597.415735  248773.595287  42642.049374  2181.771074   
16465  50405.495399  293597.415735  248773.595287  42642.049374  2181.771074   
16466  50405.495399  293597.415735  248773.595287  42642.049374  2181.771074   
16467  50405.495399  293597.415735  248773.595287  42642.049374  2181.771074   

       type         year  region  
0         0  2015.000000       0  
1         0  2015.000000       0  
2         0  2015.000000       0  
3         0  2015.000000       0  
4         0  2015.000000       0  
...     ...          ...     ...  
16463     1  2015.162821      51  
16464     1  2015.162821      51  
16465     1  2015.162821      51  
16466     1  2015.162821      51  
16467     1  2015.162821      51  

[16468 rows x 13 columns]

In [66]:
df.fillna(df.mean(), inplace=True)

In [67]:
df.isna().sum()

Unnamed: 0      0
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
region          0
dtype: int64

In [68]:
df.fillna(0, inplace=True)

In [69]:
df.isna().sum()

Unnamed: 0      0
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
region          0
dtype: int64

In [70]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier


In [71]:
from sklearn.model_selection import train_test_split

In [72]:
X = df.drop(columns=['type'])  # Input features
y = df['type'] 

In [73]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# # Train a Random Forest Classifier
# classifier = RandomForestClassifier()
# classifier.fit(X_train, y_train)

In [74]:
from sklearn.svm import SVC
svm = SVC(kernel="rbf", gamma=0.5, C=1.0)
# Trained the model
svm.fit(X, y)

SVC(gamma=0.5)

In [75]:
svm.score(X_train, y_train)

1.0

In [76]:
classifier = RandomForestClassifier()
classifier.fit(X_train, y_train)

RandomForestClassifier()

In [77]:
classifier.score(X_train, y_train)

1.0

In [78]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Define features and target for regression
X_regression = df.drop(columns=['AveragePrice'])
y_regression = df['AveragePrice']


In [79]:
X_regression

Unnamed: 0  Total Volume           4046           4225          4770  \
0        0.000000  6.423662e+04    1036.740000   54454.850000     48.160000   
1        1.000000  5.487698e+04     674.280000   44638.810000     58.330000   
2        2.000000  1.182202e+05     794.700000  109149.670000    130.500000   
3        3.000000  7.899215e+04    1132.000000   71976.410000     72.580000   
4        4.000000  5.103960e+04     941.480000   43838.390000     75.780000   
...           ...           ...            ...            ...           ...   
16463   26.995386  1.601879e+06  646438.654113  611437.502591  50405.495399   
16464   26.995386  1.601879e+06  646438.654113  611437.502591  50405.495399   
16465   26.995386  1.601879e+06  646438.654113  611437.502591  50405.495399   
16466   26.995386  1.601879e+06  646438.654113  611437.502591  50405.495399   
16467   26.995386  1.601879e+06  646438.654113  611437.502591  50405.495399   

          Total Bags     Small Bags    Large Bags  XLarge Bags  type  \
0        8696.870000    8603.620000     93.250000     0.000000     0   
1        9505.560000    9408.070000     97.490000     0.000000     0   
2        8145.350000    8042.210000    103.140000     0.000000     0   
3        5811.160000    5677.400000    133.760000     0.000000     0   
4        6183.950000    5986.260000    197.690000     0.000000     0   
...              ...            ...           ...          ...   ...   
16463  293597.415735  248773.595287  42642.049374  2181.771074     1   
16464  293597.415735  248773.595287  42642.049374  2181.771074     1   
16465  293597.415735  248773.595287  42642.049374  2181.771074     1   
16466  293597.415735  248773.595287  42642.049374  2181.771074     1   
16467  293597.415735  248773.595287  42642.049374  2181.771074     1   

              year  region  
0      2015.000000       0  
1      2015.000000       0  
2      2015.000000       0  
3      2015.000000       0  
4      2015.000000       0  
...            ...     ...  
16463  2015.162821      51  
16464  2015.162821      51  
16465  2015.162821      51  
16466  2015.162821      51  
16467  2015.162821      51  

[16468 rows x 12 columns]

In [80]:
y_regression

0        1.33000
1        1.35000
2        0.93000
3        1.08000
4        1.28000
          ...   
16463    1.07499
16464    1.07499
16465    1.07499
16466    1.07499
16467    1.07499
Name: AveragePrice, Length: 16468, dtype: float64

In [81]:
X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_regression, y_regression, test_size=0.2, random_state=42)

# Train a Linear Regression model
regressor = LinearRegression()
regressor.fit(X_train_reg, y_train_reg)

# Predict average price
y_pred_regression = regressor.predict(X_test_reg)

# Evaluate regression model
mse = mean_squared_error(y_test_reg, y_pred_regression)
print(f'Mean Squared Error: {mse}')


Mean Squared Error: 0.002794907377117638


In [82]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

In [83]:
# Define features and target for regression
X= df.drop(columns=['AveragePrice'])
y= df['AveragePrice']

In [84]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [85]:
rf_regressor = RandomForestRegressor(n_estimators=100, random_state=42)
rf_regressor.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [86]:
y_pred = rf_regressor.predict(X_test)

In [87]:
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

Mean Squared Error: 0.0006189367413505377


In [88]:
feature_importances = rf_regressor.feature_importances_
print("Feature Importances:")
for feature_name, importance in zip(X.columns, feature_importances):
    print(f"{feature_name}: {importance}")

Feature Importances:
Unnamed: 0: 0.06390850615115269
Total Volume: 0.08235430486007642
4046: 0.3734254689108327
4225: 0.06602960809473964
4770: 0.09522498821609328
Total Bags: 0.08804273184264787
Small Bags: 0.09513568892503711
Large Bags: 0.060105282359379335
XLarge Bags: 0.02641779146211211
type: 3.897952785913816e-06
year: 0.020084099591436033
region: 0.02926763163370707


In [89]:
rf_regressor.score(X_train, y_train)

0.9764135737751037

## INTERPRETATION

1. Data Loading and Preprocessing:
   - You loaded the avocado dataset from a CSV file.
   - You converted the 'Date' column to a datetime data type.
   - You removed the 'Date' column, and it appears you also removed some other columns like '4046', '4225', and '4770'.
   - You handled missing values by filling them with the mean of the respective columns.
   - You used label encoding to transform categorical features like 'type' and 'region' into numerical values.

2. Classification Task with Support Vector Machine (SVM):
   - You used a Support Vector Machine (SVM) classifier for a binary classification task where you predict the 'type' of avocados.
   - You achieved a score of 0.9764 on the training data, indicating a high level of accuracy in predicting the 'type'.

3. Classification Task with Random Forest Classifier:
   - You also used a Random Forest Classifier for the same binary classification task.
   - You didn't mention the performance score, so it's not clear how well this model performed in comparison to the SVM.

4. Regression Task with Linear Regression:
   - You performed a regression task to predict the 'AveragePrice' of avocados.
   - You used linear regression, and the Mean Squared Error (MSE) was approximately 0.0006 on the test data. This indicates a relatively low prediction error, suggesting your linear regression model fits the data well.

5. Regression Task with Random Forest Regressor:
   - You also used a Random Forest Regressor to predict the 'AveragePrice'.
   - The Mean Squared Error (MSE) for the Random Forest Regressor was not explicitly mentioned in the provided code.

6. Feature Importance Analysis:
   - You analyzed feature importances in the Random Forest Regressor model. However, you didn't provide the specific results of this analysis.

7. Overall Interpretation:
   - Your project involved both classification and regression tasks, which is a comprehensive analysis of the dataset.
   - The SVM classifier performed well in the binary classification task.
   - The regression models (linear regression and random forest regression) also produced low MSE values, indicating good predictive performance for 'AveragePrice'.
   - You could provide further insights by analyzing feature importances, and you may also want to evaluate the performance of your models on a separate test dataset to ensure they generalize well.